src/monitoring/aws/api-gateway/metrics.ts

```diff
@@ -0,0 +1,74 @@
+
+const enum Metrics {
+  FourHundredError = '4XXError',
+  FiveHundredError = '5XXError',
+  CacheHitCount = 'CacheHitCount',
+  CacheMissCount = 'CacheMissCount',
+  Count = 'Count',
+  IntegrationLatency = 'IntegrationLatency',
+  Latency = 'Latency',
+}
+
+const Namespace = 'AWS/ApiGateway';
+
+export class ApiGatewayMetricFactory {
+  metricErrors(apiName: string, stage: string, op?: WatchedOperation) {
+    return {
+      errors4XX: this.metric(Metrics.FourHundredError, apiName, stage, op).with({ label: '4XX Errors', statistic: Statistic.SUM }),
+      errors5XX: this.metric(Metrics.FiveHundredError, apiName, stage, op).with({ label: '5XX Errors', statistic: Statistic.SUM }),
+    };
+  }
+
+  metricCache(apiName: string, stage: string, op?: WatchedOperation) {
+    return {
+      hits: this.metric(Metrics.CacheHitCount, apiName, stage, op).with({ label: 'Cache Hit', statistic: Statistic.SUM }),
+      misses: this.metric(Metrics.CacheMissCount, apiName, stage, op).with({ label: 'Cache Miss', statistic: Statistic.SUM }),
+    };
+  }
+
+  metricCalls(apiName: string, stage: string, op?: WatchedOperation) {
+    return this.metric(Metrics.Count, apiName, stage, op).with({ label: 'Calls', statistic: Statistic.SUM });
+  }
+
+  metricIntegrationLatency(apiName: string, stage: string, op?: WatchedOperation) {
+    return {
+      min: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'min', statistic: Statistic.MINIMUM }),
+      avg: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'avg', statistic: Statistic.AVERAGE }),
+      p90: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p90', statistic: 'p90' }),
+      p95: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p95', statistic: 'p95' }),
+      p99: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p99', statistic: 'p99' }),
+      max: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'max', statistic: Statistic.MAXIMUM }),
+    };
+  }
+
+  metricLatency(apiName: string, stage: string, op?: WatchedOperation) {
+    return {
+      min: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'min', statistic: Statistic.MINIMUM }),
+      avg: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'avg', statistic: Statistic.AVERAGE }),
+      p90: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p90', statistic: 'p90' }),
+      p95: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p95', statistic: 'p95' }),
+      p99: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p99', statistic: 'p99' }),
+      max: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'max', statistic: Statistic.MAXIMUM }),
+    };
+  }
+
+  protected metric(metricName: Metrics, apiName: string, stage: string, op?: WatchedOperation) {
+    return new Metric({
```

Comment:
This is typically the correct way to implement metric factories. We can try and use similar pattern (of creating a single `metric` function and all other helper extend it).

Reply:
Something in that direction - I plan to create our own metric objects later, since it is important to keep track on what metric supports alarming, what period it has, what unit (which might be different than datapoint unit and might be even used to derive axis), and so on. So this is a certain compromise to save on code before this happens.